src/main/java/com/google/devtools/build/lib/pkgcache/PackageOptions.java

```diff
@@ -189,10 +189,23 @@ public ParallelismConverter() throws OptionsParsingException {
     effectTags = {OptionEffectTag.UNKNOWN},
     help =
         "Check for modifications made to the output files of a build. Consider setting "
-            + "this flag to false to see the effect on incremental build times."
+            + "this flag to false if you don't expect these files to change outside of bazel "
+            + "since it will speed up incremental build times."
   )
   public boolean checkOutputFiles;
 
+  @Option(
+    name = "experimental_check_external_files",
```

Comment:
Can we add a test (ideally a Java integration test) for this functionality?

Reply:
Sure! Can you point me at a test which is a good one to copy? I see lots of different types of ones (but none for the param which I'm copying here)

Reply:
One is here: https://github.com/bazelbuild/bazel/blob/master/src/test/java/com/google/devtools/build/lib/skyframe/OutputsInvalidationIntegrationTest.java. You should be able to do something pretty straightforward in test/.../lib/bazel/repository where you inherit from BuildIntegrationTestCase, do a trivial build with a repository, modify a file, and observe no change when you do a second build.

Reply:
Thanks for the pointers. I've been bashing my head against it for a while. I have something semi-working but it isn't registering the repo. I have this file:
```
package com.google.devtools.build.lib.bazel.repository;

import static com.google.common.truth.Truth.assertThat;

import com.google.devtools.build.lib.buildtool.BuildResult;
import com.google.devtools.build.lib.buildtool.util.BuildIntegrationTestCase;
import com.google.devtools.build.lib.testutil.MoreAsserts;
import com.google.devtools.build.lib.vfs.FileSystemUtils;
import com.google.testing.junit.testparameterinjector.TestParameterInjector;
import com.google.testing.junit.testparameterinjector.TestParameters;
import org.junit.Test;
import org.junit.runner.RunWith;

@RunWith(TestParameterInjector.class)
public class ExternalsInvalidationIntegrationTest extends BuildIntegrationTestCase {

  @TestParameters({
      "{checkExternalFiles: true}",
      "{checkExternalFiles: false}",
  })
  @Test
  public void noCheckExternalFiles_ignoresModifiedFile(boolean checkExternalFiles) throws Exception {
    addOptions("--experimental_check_external_files="+checkExternalFiles);
    write("WORKSPACE", FileSystemUtils.readLinesAsLatin1(getWorkspace().getRelative("WORKSPACE"))
        + "\n"
        + "load(\"@bazel_tools//tools/build_defs/repo:http.bzl\", \"http_archive\")\n"
            + "http_archive(\n"
            + "    name = \"rules_python\",\n"
            + "    url = \"https://github.com/bazelbuild/rules_python/releases/download/0.5.0/rules_python-0.5.0.tar.gz\",\n"
            + "    sha256 = \"cd6730ed53a002c56ce4e2f396ba3b3be262fd7cb68339f0377a45e8227fe332\",\n"
            + ")");
    write("foo/BUILD", "load(\"@rules_python//python:defs.bzl\", \"py_binary\")\n"
        + "py_binary(name = 'foo', srcs = ['main.py'])");
    write("foo/main.py", "print('hi')");
    buildTarget("//foo");
    String scanString = "About to scan skyframe graph checking for filesystem nodes of types";
    MoreAsserts.assertContainsEvent(events.collector(), scanString);

    events.collector().clear();
    writeAbsolute(getOutputBase().getRelative("external/rules_python/python/defs.bzl"), "a broken file");
    BuildResult result = buildTarget("//foo");
    assertThat(result.getSuccess()).isEqualTo(checkExternalFiles);

    MoreAsserts.assertDoesNotContainEvent(events.collector(),  scanString);
  }

}
```

but when running it I get:
```
com.google.devtools.build.lib.skyframe.BzlLoadFunction$BzlLoadFunctionException: com.google.devtools.build.lib.skyframe.BzlLoadFunction$BzlLoadFailedException: Unable to find package for @rules_python//python:defs.bzl: The repository '@rules_python' could not be resolved: Repository '@rules_python' is not defined.
```

Reply:
Next up I'll editing a `local_repository` which is pre-registered but @keith wanted that to not be skipped by this flag so I was hoping to avoid that since the semantics for that aren't locked in yet.

Reply:
Oh I see, it is all mocked out in `BazelAnalysisMock.java`

Reply:
I banged my head against this for a few more hours and got something to execute, but no matter what changing that file is observed by the cache. Maybe `local_repository` isn't cached? Anyways, this test is 100x harder to write than the original code :( 

```
package com.google.devtools.build.lib.bazel.repository;

import static com.google.common.truth.Truth.assertThat;

import com.google.devtools.build.lib.buildtool.BuildResult;
import com.google.devtools.build.lib.buildtool.util.BuildIntegrationTestCase;
import com.google.devtools.build.lib.testutil.MoreAsserts;
import com.google.devtools.build.lib.testutil.Scratch;
import com.google.devtools.build.lib.vfs.FileSystemUtils;
import com.google.devtools.build.lib.vfs.ModifiedFileSet;
import com.google.devtools.build.lib.vfs.Path;
import com.google.devtools.build.lib.vfs.Root;
import com.google.testing.junit.testparameterinjector.TestParameterInjector;
import com.google.testing.junit.testparameterinjector.TestParameters;
import java.util.logging.Logger;
import org.junit.Test;
import org.junit.runner.RunWith;

@RunWith(TestParameterInjector.class)
public class ExternalsInvalidationIntegrationTest extends BuildIntegrationTestCase {

  @TestParameters({
      "{checkExternalFiles: true}",
      "{checkExternalFiles: false}",
  })
  @Test
  public void noCheckExternalFiles_ignoresModifiedFile(boolean checkExternalFiles) throws Exception {
    addOptions("--noexperimental_check_output_files");
    addOptions("--experimental_check_external_files="+checkExternalFiles);
    write("fake_external/pkg/my_alias.bzl",
        "def my_alias(name, other):",
        "    print('my_alias ran')");
    write("fake_external/WORKSPACE", "workspace(name = 'fake_external')");
    write("fake_external/pkg/BUILD", "exports_files(['my_alias.bzl'])");
    write("WORKSPACE", FileSystemUtils.readLinesAsLatin1(getWorkspace().getRelative("WORKSPACE"))
        + "\n"
        + "local_repository(name = 'fake_external', path = 'fake_external')\n"
    );
    write("foo/BUILD", "load(\"@fake_external//pkg:my_alias.bzl\", \"my_alias\")\n"
        + "genrule(name='foo', outs=['foo.out'], cmd='echo genrule ran; touch $@')\n"
        + "my_alias(name='bar', other='foo')");
    buildTarget("//foo");

    writeAbsolute(getOutputBase().getRelative("external/fake_external/pkg/my_alias.bzl"), "a broken file");
    try {
      buildTarget("//foo");
      if (!checkExternalFiles) {
        throw new Exception("Should have raised");
      }
    } catch (Exception e) {
      if (checkExternalFiles) {
        throw new Exception("Should not have raised");
      }
    }
  }

}
```